### ISSUE TITLE
Offset paging: do not materialize items while there's no "nodes" in request

### ISSUE THREAD

**Issue by ademchenko** [ChilliCream/graphql-platform#2520]
I've done some review of OffsetPaging made in version 11. The good point is that you have made an optimization regarding not loading totalCount once it's not requested. Actually, we use HotChocolate on millions of rows and the issue with "totalCount" is what we also faced. So, in our local (based on version 10) OffsetPaging implementation, we have also done that optimization. Though it is a little bit easier - we don't set up getTotalCount fetching function at all if "totalCount" is not requested because now I don't see a chance for "getTotalCount" to be called at all when "totalCount" is not requested. If you could explain that case, I would really appreciate it.

But one further optimization we had to make is a similar to "totalCount"  approach in regards to "nodes", and since we are keen on upgrading to a new version 11.0 and to remove our local OffsetPaging implementation, we kindly ask you to make the requested optimization as well.

Our case in details:

Some of our tables are very heavy. We have millions and millions of rows and have to deal with them. There are so many data that we need to use "Map-Reduce" algorithm to process such a number of rows.

It's impossible to load all the data at once, so we do that loading data by chunks. The size of chunk depends not only on the total count but also on some technical details of "Map-Reduce" engine (the number of working reducers, etc., etc.). To calculate the size of the chunk, we, first of all, need to get "totalCount". So, we use the following approach:
1. The first request to graphql server is made only for "totalCount" but **no nodes**
2. After we got "totalCount" fetched we can calculate the chunk size
3. Having the chunk size we make a series of requests  (they even might be done from the different physical nodes) for "nodes" but **no "totalCount"**
---
4. For some reasons it's not an issue for us that the actual total count can be changed across the requests for nodes.

So as you could see we use both optimizations (which you have already made and the proposed one) heavily and so kindly request to implement that feature request as well.

**Comment by michaelstaib** [comment]
HI @ademchenko,

are you using IQueryable for data or some other means. We have abstracted paging and you basically can create your own paging provider now. Just looked at the default implementation and I can rework it a little so that it can be easily inherited and extended.

Do you think that is what you want?

A provider has just to be added to the dependency injection and is used by the paging middleware.

**Comment by ademchenko** [comment]
Hi, @michaelstaib . Thanks for your quick response. 

Yes, we use IQueryable which is connected to the database (so not materialized).

@michaelstaib , I really support all the activity regarding increasing extensibility of HotChocolate, but concerning our case, I think it's more appropriate to add that feature inside the Core.

Let's see on it from the logical point of view:

1. We have just two root output fields: "totalCount" and "nodes"
2. Each field always requires a separate query to the database 
3. Queries are interconnected (they have common filter and order condition), so we can assume that if one of the queries is heavy than the second one is probably heavy as well
4. According to the syntax it is not prohibited to request only one field (any of "totalCount" and "nodes") from two
5. I gave a real example when it's needed to request "totalCount" and "nodes" separately

According to 1-5, in terms of usage "totalCount" and "nodes" have absolutely the same behaviour. It means they have to be treated in the same way. But your implementation doesn't actually do that. You have made a good optimization job regarding "totalCount" but nothing regarding "nodes".

@michaelstaib , given the above, I believe the best way would be to go one step further and apply the "totalCount" optimization to "nodes" as well (inside the Core implementation).

**Comment by michaelstaib** [comment]
Do you want to do a pr?

**Comment by ademchenko** [comment]
Sure, I will.

**Comment by PascalSenn** [comment]
@ademchenko  i just refactored a fair bit of the paging stuff. So maybe we want to wait till this is merged. because there is no way we would get this together when we both did major changes

**Comment by michaelstaib** [comment]
@PascalSenn is it ready for review?

**Comment by ademchenko** [comment]
@michaelstaib, @PascalSenn, could you please answer on that:

>  we don't set up getTotalCount fetching function at all if "totalCount" is not requested because now I don't see a chance for "getTotalCount" to be called at all when "totalCount" is not requested. If you could explain that case, I would really appreciate it.

Why did you decide to leave the possibility of a deferred call, which part of the system could potentially make that call if the output field is not presented? Some custom middleware could add the field to output?

**Comment by michaelstaib** [comment]
I will review the code and give you an answer on this.

**Comment by ademchenko** [comment]
@michaelstaib , when will that possible to start?

**Comment by PascalSenn** [comment]
@ademchenko  it's merged

**Comment by michaelstaib** [comment]
Regarding the deferred call. So, the deferred total count is meant for the execution engine. If the total count is enabled then the execution engine will invoke the resolver for it. If not it will not be invoked. If the handler has a better way to get the total count on invoking the query it can do so and the total count will essentially just be a cached value.

**Comment by michaelstaib** [comment]
Can we close this issue?

**Comment by ademchenko** [comment]
@michaelstaib , yes, but why do we need to defer that call? Why isn't it suitable to make that call right in the place?

@michaelstaib , I would ask you not to close that issue if that's possible. Without that optimization we won't be able to move to version 11 since we are handling millions and millions of items through the IQueryable and even "totalCount" takes quite a bit of processing time. I will make that PR once I understand clearly with the deferred call.

**Comment by michaelstaib** [comment]
>  yes, but why do we need to defer that call? Why isn't it suitable to make that call right in the place?

because it might never happen and also can now be done in parallel. The execution engine now can do both calls in one go since they are represented by two resolvers. But you do not need to handle it this way. You also can do it in one call instead of deferring it.

**Comment by michaelstaib** [comment]
>  would ask you not to close ...

No worries ... just checking in. If it is not done we can leave it open.

**Comment by gojanpaolo** [comment]
Quick question, why does `UsePaging` apply `nodes` while `UseOffsetPaging` apply `items`? Wouldn't it be better to choose the same default name for consistency? Thanks!

**Comment by ademchenko** [comment]
@gojanpaolo, the "node"  was in my commit I've made as a pull request. But, it looks like the team decided to split the conceptions. To be honest, I'm not sure why. Maybe, @michaelstaib  could answer.

**Comment by ademchenko** [comment]
This is critical issue which really blocks us from using the API because it has performance issues.

**Comment by ademchenko** [comment]
Another serious issue is IPage.HasNextPage.


In order to decide if there's the next page or not, the _"take"_ plus **1 additional item** is materialized for **each retrieved page**. But, team, please, don't take into account only the UI scenario when some user requests pages into its grid. In the UI case, there's no high productivity, and it's more or less okay to retrieve several extra items. But, for example, in our case, the backend uses the graphql API which processes hundreds of millions of rows chunk by chunk. It's extremely important for us to turn off the "HasNextPage" option. With offset paging, even for UI grids, there's no need for that option. Just recall how that really proceeded - you see the list of numbers of the pages from the first to the last one. There's nothing to do with HasNextPage even in the UI scenario.


Please, implement the ability to turn off the  _HasPreviousPage_ and _HasNextPage_ functionality.

**Comment by michaelstaib** [comment]
Sorry for closing this.

**Comment by ademchenko** [comment]
Hello, team. 

This is a very important and blocking issue for some of our scenarios. Is there an estimate of when you can start working on that issue? If you need I can help. Just give me ideas about which solution you would like me to implement in order to follow your architecture plan.

**Comment by ademchenko** [comment]
Hello, team!

Any news on that? We are still facing performance issues because of non-optimal behaviour described in that task.

**Comment by michaelstaib** [comment]
We will keep offset pagination as is around but will not invest in it anymore. The reason for this is that we have invested heavily in cursor pagination and with 15.1 it supports all use cases of offset pagination with better performance and better dataset stability.

However for users that currently use offset pagination we will keep it around in order to not break your schemas.
